experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,957 @@
+  // get the value of the wire
+  virtual std::any getData() = 0;
+  // print function for convenience
+  virtual void printValue() = 0;
+  virtual ~Wire() = default;
+};
+
+/// represent the concrete port as a smart pointer to the memory
+using Port = std::shared_ptr<Wire>;
+
+/// result = a && b in terms of ports
+void conjunction(Port &a, Port &b, Port &result) {
+  if (a->hasValue() || b->hasValue()) {
+    if (a->hasValue() && b->hasValue()) {
+      result->updateValue(std::any_cast<bool>(a->getData()) &&
+                          std::any_cast<bool>(b->getData()));
+    } else if (a->hasValue()) {
+      if (!std::any_cast<bool>(a->getData())) {
+        result->updateValue(false);
+      } else {
+        result = b;
+      }
+
+    } else {
+      if (!std::any_cast<bool>(b->getData())) {
+        result->updateValue(false);
+      } else {
+        result = a;
+      }
+    }
+  }
+}
+
+/// result = a || b in terms of ports
+void disjunction(Port &a, Port &b, Port &result) {
+  if (a->hasValue() || b->hasValue()) {
+    if (a->hasValue() && b->hasValue()) {
+      result->updateValue(std::any_cast<bool>(a->getData()) ||
+                          std::any_cast<bool>(b->getData()));
+    } else if (a->hasValue()) {
+      if (std::any_cast<bool>(a->getData())) {
+        result->updateValue(true);
+      } else {
+        result = b;
+      }
+
+    } else {
+      if (std::any_cast<bool>(b->getData())) {
+        result->updateValue(true);
+      } else {
+        result = a;
+      }
+    }
+  }
+}
+
+/// result = ~a in terms of ports
+void negation(Port &a, Port &result) {
+  if (a->hasValue()) {
+    result->updateValue(!std::any_cast<bool>(a->getData()));
+  }
+}
+
+/// convert vector<bool> to integer
+uint64_t vec2num(const std::vector<bool> &v, unsigned bitwidth) {
+  uint64_t result = 0;
+  uint64_t tmp;
+  for (unsigned i = 0; i < bitwidth; ++i) {
+    tmp = v[i];
+    result |= tmp << (bitwidth - i - 1);
+  }
+  return result;
+}
+
+/// convert integer to vector<bool>
+std::vector<bool> num2vec(uint64_t num, unsigned bitwidth) {
+  std::vector<bool> tempVec;
+  while (num) {
+    tempVec.push_back(num & 1u);
+    num >>= 1;
+  }
+  if (bitwidth > tempVec.size()) {
+    for (unsigned i = tempVec.size(); i < bitwidth; ++i) {
+      tempVec.push_back(false);
+    }
+  } else {
+    for (unsigned i = bitwidth; i < tempVec.size(); ++i) {
+      tempVec.pop_back();
+    }
+  }
+  std::reverse(tempVec.begin(), tempVec.end());
+  return tempVec;
+}
+
+/// represents the signal
+/// in terms of VHDL : std_logic
+class Signal : public Wire {
+  std::optional<bool> signal;
+
+public:
+  Signal(std::optional<bool> s = {}) : signal(s) {}
+  bool hasValue() override { return signal.has_value(); }
+  void updateValue(std::any val) override { signal = std::any_cast<bool>(val); }
+  void printValue() override {
+    if (hasValue()) {
+      llvm::outs() << signal.value();
+    } else {
+      llvm::outs() << "U";
```

Comment:
Quoting the spec

> Non-binary states. We do not care care to model anything beyond the basic two-state logic (0 and 1). For example, we will not handle undefined or high-impedance states.

We don't even want it to be possible to model undefined values, so in no cases should we have any undefined state.